stablehlo/reference/Element.cpp

```diff
@@ -173,6 +173,221 @@ bool areApproximatelyEqual(APFloat f, APFloat g) {
+    applyIntOrFloatConversion(el.getIntegerValue());
+    return results;
+  }
+
+  if (isSupportedFloatType(operandType)) {
+    applyIntOrFloatConversion(el.getFloatValue().bitcastToAPInt());
+    return results;
+  }
+
+  if (isSupportedComplexType(operandType) && isSupportedComplexType(type)) {
+    // splits operand into 'resultBitWidth' / 'operandBitWidth' elements,
+    // applies conversion to type 'type'. It converts two elements at a
+    // time, one for the real part and the other for the imaginary part, and
+    // appends them to 'results'.
+    auto applyComplexConversion = [&](APInt mask) {
+      auto baseType = type.cast<ComplexType>().getElementType();
+      for (auto i = 0; i < operandBitWidth; i += resultBitWidth) {
+        auto binaryReal = leftShiftAndMask(
+            APInt(resultBitWidth, 0), resultBitWidth,
+            mask.extractBits(resultBitWidth, i).zext(resultBitWidth));
+        i += resultBitWidth;
+        auto binaryImag = leftShiftAndMask(
+            APInt(resultBitWidth, 0), resultBitWidth,
+            mask.extractBits(resultBitWidth, i).zext(resultBitWidth));
+        results.push_back(Element(
+            type, std::complex<APFloat>(
+                      APFloat(baseType.cast<FloatType>().getFloatSemantics(),
+                              binaryReal),
+                      APFloat(baseType.cast<FloatType>().getFloatSemantics(),
+                              binaryImag))));
+      }
+    };
+    applyComplexConversion(el.getComplexValue().real().bitcastToAPInt());
+    applyComplexConversion(el.getComplexValue().imag().bitcastToAPInt());
+    return results;
+  }
+
+  report_fatal_error(invalidArgument("Unsupported conversion from %s to %s",
+                                     debugString(operandType).c_str(),
+                                     debugString(type).c_str()));
+}
+
+// Returns bitcast converted Element object with the same bitwidth.
+Element bitcastConvertOneToOne(Type type, const Element &el) {
+  auto operandType = el.getType();
+  int64_t operandBitWidth = isSupportedComplexType(operandType)
+                                ? operandType.cast<ComplexType>()
+                                      .getElementType()
+                                      .cast<FloatType>()
+                                      .getIntOrFloatBitWidth()
+                                : operandType.getIntOrFloatBitWidth();
+  int64_t resultBitWidth = isSupportedComplexType(type)
+                               ? type.cast<ComplexType>()
+                                     .getElementType()
+                                     .cast<FloatType>()
+                                     .getIntOrFloatBitWidth()
+                               : type.getIntOrFloatBitWidth();
+  if (isSupportedBooleanType(operandType) && isSupportedBooleanType(type))
+    return el;
+
+  if (isSupportedIntegerType(operandType)) {
+    if (isSupportedIntegerType(type)) return el;
```

Comment:
It seems it may not work if operandType is `in` and type is `uin`. Also, we should add a test case for that as well.

Reply:
Nice finding! This works because the underlying storage does not change and APInt currently does not distinguish between signed/unsigned. However, we plan to distinguish them separately in #1469 so we can address them together. I'll make a note on this.

Reply:
>> The underlying element is of different type but since the tensor type is correct, the test passes.
That is right. However with the changes proposed in https://github.com/ghpvnist/stablehlo/pull/8 we may need to return the operand value `el`. 

In any case, distinguish them separately in https://github.com/openxla/stablehlo/issues/1469 make sense. 

Marking it resolved.